File: solutions/1300-1399/1397.find-all-good-strings/Solution.java
public class Solution {
    private int m, n, M = 1000000007;
    private char[] s1, s2, evil;
    private int[][][][] dp;
    private int[] pi, p;
    
    public int findGoodStrings(int n, String s1, String s2, String evil) {
        this.s1 = s1.toCharArray();
        this.s2 = s2.toCharArray();
        this.evil = evil.toCharArray();
        this.m = evil.length();
        this.n = n;
        dp = new int[n][m+1][2][2];
        for (int[][][] d : dp)
            for (int[][] e : d)
                for (int[] f : e) Arrays.fill(f, -1);

        pi = new int[m];
        char[] pat = evil.toCharArray();

        int length = 0, i = 1;
        while (i < m) {
            if (pat[i] == pat[length]) {
                pi[i++] = ++length;
            } else if (length > 0) {
                length = pi[length - 1];
            } else {
                pi[i++] = 0;
            }
        }

        this.p = new int[n+1];
        p[0] = 1;
        for (i = 1; i <= n; i++) {
            p[i] = 26L * p[i-1] % M;
        }

        return f(0, 0, 1, 1);
    }
    
    private int f(int i, int j, int b1, int b2) {
        if (j == m) return 0;
        if (i == n) return 1;
        if (dp[i][j][b1][b2] != -1) return dp[i][j][b1][b2];

        long res = 0;
        int lo = b1 > 0 ? s1[i] - 'a' : 0, hi = b2 > 0 ? s2[i] - 'a' : 25;
        for (int c = lo; c <= hi; c++) {
            int ni = i+1, nj = j;
            while (nj > 0 && evil[nj] != c + 'a') {
                nj = pi[nj-1];
            }
            if (c == evil[nj] - 'a') nj++;
            res = (res + f(ni, nj, b1 & (c == lo ? 1 : 0), b2 & (c == hi ? 1 : 0))) % M;
        }

        return dp[i][j][b1][b2] = (int) res;
    }
}

File: solutions/1300-1399/1397.find-all-good-strings/Solution.cpp
class Solution {
public:
    int findGoodStrings(int n, string s1, string s2, string evil) {
      // Initialize the dp array
        dp.resize(n, vector<vector<vector<int>>>(evil.length(), vector<vector<int>>(2, vector<int>(2, -1))));
        nextMatchedCount.resize(evil.length(), vector<int>(26, -1));
        
        return find(s1, s2, evil, 0, 0, true, true, getLPS(evil));
    }
private:
    static constexpr int kMod = 1'000'000'007;
    vector<vector<vector<vector<int>>>> dp;
    vector<vector<int>> nextMatchedCount;
    
    int find(const string& s1, const string& s2, const string& evil, int i, int matchedEvilCount, bool isS1Prefix, bool isS2Prefix, const vector<int>& evilLPS) {
        if (matchedEvilCount == evil.length()) return 0;
        if (i == s1.length()) return 1;

        int& ans = dp[i][matchedEvilCount][isS1Prefix][isS2Prefix];
        if (ans != -1) return ans;

        ans = 0;
        const char minChar = isS1Prefix ? s1[i] : 'a';
        const char maxChar = isS2Prefix ? s2[i] : 'z';
        
        for (char c = minChar; c <= maxChar; ++c) {
            const int nextMatchedEvilCount = getNextMatchedEvilCount(evil, matchedEvilCount, c, evilLPS);
            ans += find(s1, s2, evil, i + 1, nextMatchedEvilCount, isS1Prefix && c == s1[i], isS2Prefix && c == s2[i], evilLPS);
            ans %= kMod;
        }
        return ans;
    }

    // Get Longest Prefix also Suffix
    vector<int> getLPS(const string& s) {
        vector<int> lps(s.length());
        for (int i = 1, j = 0; i < s.length(); ++i) {
            while (j > 0 && s[j] != s[i]) j = lps[j - 1];
            if (s[i] == s[j]) lps[i] = ++j;
        }
        return lps;
    }

    int getNextMatchedEvilCount(const string& evil, int j, char currChar, const vector<int>& evilLPS) {
        int& ans = nextMatchedCount[j][currChar - 'a'];

        if (ans != -1) return ans;
        while (j > 0 && evil[j] != currChar) j = evilLPS[j - 1];
        return ans = (evil[j] == currChar ? j + 1 : j);
    }
};